src/NuGet.Protocol.Core.v3/HttpHandlerResourceV3Provider.cs

```diff
@@ -104,6 +104,20 @@ private class ProxyWebRequestHandler : WebRequestHandler
                             HttpHandlerResourceV3.ProxyPassed(Proxy);
                         }
 
+                        if (response.StatusCode == HttpStatusCode.Unauthorized)
+                        {
+                            if (HttpHandlerResourceV3.PromptForProxyCredentials == null)
+                            {
+                                // Fail this operation.
+                                response.EnsureSuccessStatusCode();
+                            }
+                            else
+                            {
+                                // PromptForProxyCredentials is a misnomer. It can prompt for proxy credentials or request credentials.
+                                Credentials = HttpHandlerResourceV3.PromptForProxyCredentials(request.RequestUri, /* webProxy */ null);
```

Comment:
I think it's much better to add a new delegate HttpHandlerResourceV3.PromptForCredentials(). PromptForProxyCredentails() does exactly what's it's named: asking for Proxy credentials. Thus, it also syncs with the v2 ProxyCache.

So similarly, the new PromptForCredentials() would need to sync with v2 CredentialStore.

you also need to add a new delegate, say, HttpHandlerResourceV3.CredentialPassed, which, similar to HttpHandlerResourceV3.ProxyPassed, syncs v3 credentials with v2.